Lib/code.py

```diff
@@ -140,32 +132,38 @@ def showtraceback(self, **kwargs):
 
         """
         colorize = kwargs.pop('colorize', False)
-        sys.last_type, sys.last_value, last_tb = ei = sys.exc_info()
-        sys.last_traceback = last_tb
-        sys.last_exc = ei[1]
         try:
-            if sys.excepthook is sys.__excepthook__:
-                lines = traceback.format_exception(ei[0], ei[1], last_tb.tb_next, colorize=colorize)
-                self.write(''.join(lines))
-            else:
-                # If someone has set sys.excepthook, we let that take precedence
-                # over self.write
-                self._call_excepthook(ei[0], ei[1], last_tb)
+            typ, value, tb = sys.exc_info()
+            # Remove internal frames (in modules code and codeop) from traceback.
+            while tb and tb.tb_frame.f_globals.get('__name__') in ('code', 'codeop'):
```

Comment:
unittest's `_clean_tracebacks()` also looks at the chained exceptions.

Maybe it makes sense to implement this in the traceback module, and reuse?

Reply:
Actually, even this loop is an overpower on one side and not enough on other side.  I initially added it to handle both compile-time (this is why `copeop`) and run-time errors. But then I have found that compile-time errors should not have a traceback at all.

Now I see that we should just remove one level from a traceback -- the one that refers to the code calls `exec()`. This is also necessary for IDLE which overrides `runcode()` -- in this case the top frame is in the `idlelib.pyshell` module.

This will not work if the code calls `exec()` not directly but via wrapper. Ideally we should find the frame that calls `exec()` (or `eval()`) and drop it and above frames, but there is no easy way to do this. It would help if `exec()` and `eval()` add a special marker in the traceback, but this would be a new feature.